src/lib/GenericTraversal.hs

```diff
@@ -22,28 +22,46 @@ import MTL1
 
 import LabeledItems
 
-liftGenericTraverserM :: EnvReader m => s n -> GenericTraverserM s n n a -> m n (a, s n)
+liftGenericTraverserM :: EnvReader m => s n -> GenericTraverserM UnitB s n n a -> m n (a, s n)
 liftGenericTraverserM s m =
-  liftBuilder $ runStateT1 (runSubstReaderT idSubst $ runGenericTraverserM' m) s
+  liftM runHardFail $ liftDoubleBuilderTNoEmissions $
+    runStateT1 (runSubstReaderT idSubst $ runGenericTraverserM' m) s
 {-# INLINE liftGenericTraverserM #-}
 
-newtype GenericTraverserM s i o a =
+liftGenericTraverserMTopEmissions
+  :: (EnvReader m, SinkableE e, SubstE Name e, SinkableE s, SubstE Name s)  => s n
+  -> (forall l. DExt n l => GenericTraverserM TopEnvFrag s l l (e l))
+  -> m n (Abs TopEnvFrag (PairE e s) n)
+liftGenericTraverserMTopEmissions s m =
+  liftM runHardFail $ liftDoubleBuilderT do
+    (e, s') <- runStateT1 (runSubstReaderT idSubst $ runGenericTraverserM' m) (sink s)
+    return $ PairE e s'
+{-# INLINE liftGenericTraverserMTopEmissions #-}
+
+newtype GenericTraverserM f s i o a =
```

Comment:
I'm not sure how I feel about this extra parameter. I am a little afraid that it will inhibit GHC optimizations and I want to double check that before we merge this. It also seems slightly overkill just for this one Ix hoisting pass. I thought we were going to do it inside simplification, where we don't use the generic traversal, so we wouldn't need this.

Reply:
I can try moving it to simplification. As I mentioned above the only reason was that I didn't want to turn all the `substM ty` lines in Simplify to `simplifyType ty` or `simplifyAtom ty` or something. I agree that we should check whether it inhibits optimizations. Even if we don't need it for ix hoisting (say, because we move that to simplification), I suspect we'll want to have the freedom to emit top-level emissions in other places too. But we can delete it until we need it if we don't need it now.